src/AppInstallerRepositoryCore/Microsoft/SQLiteStorageBase.h

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include "SQLiteWrapper.h"
+#include "Microsoft/Schema/Version.h"
+#include <winget/ManagedFile.h>
+#include <AppInstallerVersions.h>
+
+#include <mutex>
+
+namespace AppInstaller::Repository::Microsoft
+{
+    struct SQLiteStorageBase
+    {
+        // The disposition for opening the index.
+        enum class OpenDisposition
+        {
+            // Open for read only.
+            Read,
+            // Open for read and write.
+            ReadWrite,
+            // The database will not change while in use; open for immutable read.
+            Immutable,
+        };
+
+        static char const* const GetOpenDispositionString(SQLiteStorageBase::OpenDisposition disposition)
+        {
+            switch (disposition)
+            {
+            case AppInstaller::Repository::Microsoft::SQLiteStorageBase::OpenDisposition::Read:
+                return "Read";
+            case AppInstaller::Repository::Microsoft::SQLiteStorageBase::OpenDisposition::ReadWrite:
+                return "ReadWrite";
+            case AppInstaller::Repository::Microsoft::SQLiteStorageBase::OpenDisposition::Immutable:
+                return "ImmutableRead";
+            default:
+                return "Unknown";
+            }
+        }
+
+        template<typename T>
+        static T Open(const std::string& filePath, OpenDisposition disposition, Utility::ManagedFile&& indexFile = {})

```

Comment:
Please don't use this pattern, especially while leaving `CreateNew` as a static member of the derived types.

The previous `Open` should become a constructor on `SQLiteStorageBase`, then the derived types should `using SQLiteStorageBase;` to promote the constructors to their type.  Then their `Open` becomes simply a passthrough:
```
Open(params)
{
  return { params };
}
```

Reply:
I created a new SQLiteStorageBase constructor that performs the previous "Open" function and also added new constructors for both SQLiteIndex and PortableIndex that inherit from this new base "Open" constructor. This allows me to call the "create interface" function and set it.